File: tests/leetcode/x/kotlin/0174.kt
fun solve(dungeon: List<List<Int>>): Int {
    val cols = dungeon[0].size
    val inf = 1_000_000_000
    val dp = MutableList(cols + 1) { inf }
    dp[cols - 1] = 1
    for (i in dungeon.size - 1 downTo 0) {
        for (j in cols - 1 downTo 0) {
            val need = minOf(dp[j], dp[j + 1]) - dungeon[i][j]
            dp[j] = if (need <= 1) 1 else need
        }
    }
    return dp[0]
}

fun main() {
    val toks = generateSequence { readLine() }.flatMap { it.trim().split(Regex("\\s+")).asSequence() }.filter { it.isNotEmpty() }.toList()
    if (toks.isEmpty()) return
    var idx = 0
    val t = toks[idx++].toInt()
    val out = ArrayList<String>()
    repeat(t) {
        val rows = toks[idx++].toInt()
        val cols = toks[idx++].toInt()
        val dungeon = ArrayList<List<Int>>()
        repeat(rows) {
            val row = ArrayList<Int>()
            repeat(cols) { row.add(toks[idx++].toInt()) }
            dungeon.add(row)
        }
        out.add(solve(dungeon).toString())
    }
    print(out.joinToString("\n"))
}

File: tests/leetcode/x/dart/0174.dart
import 'dart:io';

int solve(List<List<int>> dungeon) {
  final cols = dungeon[0].length;
  final dp = List<int>.filled(cols + 1, 1 << 60);
  dp[cols - 1] = 1;
  for (var i = dungeon.length - 1; i >= 0; i--) {
    for (var j = cols - 1; j >= 0; j--) {
      final need = (dp[j] < dp[j + 1] ? dp[j] : dp[j + 1]) - dungeon[i][j];
      dp[j] = need <= 1 ? 1 : need;
    }
  }
  return dp[0];
}

void main() {
  final toks = <String>[];
  String? line;
  while ((line = stdin.readLineSync()) != null) {
    for (final p in line!.trim().split(RegExp(r'\s+'))) {
      if (p.isNotEmpty) toks.add(p);
    }
  }
  if (toks.isEmpty) return;
  var idx = 0;
  final t = int.parse(toks[idx++]);
  final out = <String>[];
  for (var tc = 0; tc < t; tc++) {
    final rows = int.parse(toks[idx++]);
    final cols = int.parse(toks[idx++]);
    final dungeon = <List<int>>[];
    for (var i = 0; i < rows; i++) {
      final row = <int>[];
      for (var j = 0; j < cols; j++) row.add(int.parse(toks[idx++]));
      dungeon.add(row);
    }
    out.add('${solve(dungeon)}');
  }
  stdout.write(out.join('\n'));
}
